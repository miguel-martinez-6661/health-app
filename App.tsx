import React, {useCallback, useEffect, useState} from 'react';
import {Linking, SafeAreaView, ScrollView} from 'react-native';
import {HealthState} from './src/interface/health.interface';
import {RequestPermission} from './src/components/request-permission';
import {Header} from './src/components/header';
// @ts-ignore
import {initHealth, getHealthSteps} from './src/utils/health-utils';
import {HealthStepStatus} from './src/components/health-status';

const App = () => {
  const [state, setState] = useState<HealthState>();

  const checkAppPermissions = useCallback(async () => {
    Linking.openSettings();
  }, []);

  const getHealthStatus = useCallback(async () => {
    await initHealth();
    const result = await getHealthSteps();
    setState(result);
  }, []);

  useEffect(() => {
    getHealthStatus();
  }, [getHealthStatus]);

  return (
    <SafeAreaView className="bg-white h-full">
      <Header />
      <ScrollView>
        {state?.permissionGranted ? (
          <HealthStepStatus>
            <HealthStepStatus.Title>Steps</HealthStepStatus.Title>
            <HealthStepStatus.Value description="steps">
              {state?.stepCount || 0}
            </HealthStepStatus.Value>
          </HealthStepStatus>
        ) : (
          <RequestPermission onPress={checkAppPermissions} />
        )}
      </ScrollView>
    </SafeAreaView>
  );
};

export default App;
