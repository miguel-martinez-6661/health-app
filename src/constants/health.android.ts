import {HealthState} from '../interface/health.interface';
import {Scopes, BucketUnit} from 'react-native-google-fit';

export const initialState: HealthState = {
  permissionGranted: false,
  stepCount: 0,
};

export const options = {
  scopes: [Scopes.FITNESS_ACTIVITY_READ, Scopes.FITNESS_BODY_READ],
};

export const googleFitOptions = {
  startDate: '2017-01-01T00:00:17.971Z',
  endDate: new Date().toISOString(),
  bucketUnit: BucketUnit.DAY,
  bucketInterval: 1,
};
