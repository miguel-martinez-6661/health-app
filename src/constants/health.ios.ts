import AppleHealthKit, {HealthKitPermissions} from 'react-native-health';

export const options = {
  date: new Date().toISOString(), // optional; default now
  includeManuallyAdded: false, // optional: default true
};

/* Permission options */
export const permissions = {
  permissions: {
    read: [AppleHealthKit.Constants.Permissions.HeartRate],
    write: [AppleHealthKit.Constants.Permissions.Steps],
  },
} as HealthKitPermissions;
