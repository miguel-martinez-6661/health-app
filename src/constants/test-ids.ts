export const TestIDs = {
  HEADER: 'HEADER',
  HEADER_GREETINGS: 'HEADER_GREETINGS',
  HEADER_TITLE: 'HEADER_TITLE',

  HEALTH_STEP_SECTION: 'HEALTH_STEP_SECTION',
  HEALTH_STEP_NAME: 'HEALTH_STEP_NAME',
  HEALTH_STEP_VALUE: 'HEALTH_STEP_VALUE',
  HEALTH_STEP_DESCRIPTION: 'HEALTH_STEP_DESCRIPTION',

  REQUEST_PERMISSION_SECTION: 'REQUEST_PERMISSION_SECTION',
  REQUEST_PERMISSION_BUTTON: 'REQUEST_PERMISSION_BUTTON',
};
