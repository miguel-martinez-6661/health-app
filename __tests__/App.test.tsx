import React from 'react';
import App from '../App';
import {render, waitFor} from '@testing-library/react-native';
import {TestIDs} from '../src/constants/test-ids';

jest.useFakeTimers();
jest.mock('react-native-health');
jest.mock('../src/utils/health-utils', () => ({
  initHealth: jest.fn(),
  getHealthSteps: jest.fn().mockResolvedValue(() => ({
    permissionGranted: true,
    stepCount: 1000,
  })),
}));

describe('<App /> Testing', () => {
  it('should render the App component', () => {
    const {toJSON} = render(<App />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render Apple Health | Google Fit data', async () => {
    const {getByTestId} = render(<App />);
    const requestPermissionComponent = getByTestId(
      TestIDs.REQUEST_PERMISSION_SECTION,
    );
    expect(requestPermissionComponent).toBeDefined();
    await waitFor(() => {
      const stepComponent = getByTestId(TestIDs.HEALTH_STEP_SECTION);
      expect(stepComponent).toBeDefined();
    });
  });
});
