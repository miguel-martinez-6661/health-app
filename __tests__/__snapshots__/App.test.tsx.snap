// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Testing should render the App component 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "height": "100%",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
    testID="HEADER"
  >
    <Text
      style={
        [
          {
            "fontSize": 18,
            "lineHeight": 28,
          },
        ]
      }
      testID="HEADER_GREETINGS"
    >
      Hello, Miguel
    </Text>
    <Text
      style={
        [
          {
            "fontWeight": "700",
          },
          {
            "fontSize": 30,
            "lineHeight": 36,
          },
        ]
      }
      testID="HEADER_TITLE"
    >
      Activities
    </Text>
  </View>
  <RCTScrollView>
    <View>
      <View
        style={
          [
            {
              "backgroundColor": "#60a5fa",
            },
            {
              "marginBottom": 8,
              "marginLeft": 8,
              "marginRight": 8,
              "marginTop": 8,
            },
            {
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
            },
            {
              "borderBottomLeftRadius": 8,
              "borderBottomRightRadius": 8,
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
            },
          ]
        }
        testID="REQUEST_PERMISSION_SECTION"
      >
        <Text
          style={
            [
              {
                "fontSize": 18,
                "lineHeight": 28,
              },
            ]
          }
        >
          To start tracking your Health info, you should grant permissions
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#1d4ed8",
              "borderBottomLeftRadius": 8,
              "borderBottomRightRadius": 8,
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
              "marginTop": 16,
              "opacity": 1,
              "paddingBottom": 8,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
            }
          }
          testID="REQUEST_PERMISSION_BUTTON"
        >
          <Text
            style={
              [
                {
                  "color": "#fff",
                },
                {
                  "fontSize": 18,
                  "lineHeight": 28,
                },
              ]
            }
          >
            Grant Permissions
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
